ocf_datapipes/select/select_spatial_slice.py

```diff
@@ -21,7 +21,249 @@
+
+
+def _get_idx_of_pixel_closest_to_poi_geostationary(
+    xr_data: xr.DataArray,
+    center_osgb: Location,
+) -> Location:
+    """
+    Return x and y index location of pixel at center of region of interest.
+
+    Args:
+        xr_data: Xarray dataset
+        center_osgb: Center in OSGB coordinates
+
+    Returns:
+        Location for the center pixel in geostationary coordinates
+    """
+
+    xr_coords, xr_x_dim, xr_y_dim = spatial_coord_type(xr_data)
+
+    x, y = osgb_to_geostationary_area_coords(x=center_osgb.x, y=center_osgb.y, xr_data=xr_data)
+    center_geostationary = Location(x=x, y=y, coordinate_system="geostationary")
+
+    # Check that the requested point lies within the data
+    assert xr_data[xr_x_dim].min() < x < xr_data[xr_x_dim].max()
+    assert xr_data[xr_y_dim].min() < y < xr_data[xr_y_dim].max()
+
+    # Get the index into x and y nearest to x_center_geostationary and y_center_geostationary:
+    x_index_at_center = searchsorted(
+        xr_data[xr_x_dim].values, center_geostationary.x, assume_ascending=True
+    )
+
+    # y_geostationary is in descending order:
+    y_index_at_center = searchsorted(
+        xr_data[xr_y_dim].values, center_geostationary.y, assume_ascending=False
+    )
+
+    return Location(x=x_index_at_center, y=y_index_at_center, coordinate_system="idx")
+
+
+def _get_points_from_unstructured_grids(
+    xr_data: xr.DataArray,
+    location: Location,
+    location_idx_name: str = "values",
+    num_points: int = 1,
+):
+    """
+    Get the closest points from an unstructured grid (i.e. Icosahedral grid)
+
+    This is primarily used for the Icosahedral grid, which is not a regular grid,
+     and so is not an image
+
+    Args:
+        xr_data: Xarray dataset
+        location: Location of center point
+        location_idx_name: Name of the index values dimension
+            (i.e. where we index into to get the lat/lon for that point)
+        num_points: Number of points to return (should be width * height)
+
+    Returns:
+        The closest points from the grid
+    """
+    xr_coords, xr_x_dim, xr_y_dim = spatial_coord_type(xr_data)
+    assert xr_coords == "lon_lat"
+
+    # Check if need to convert from different coordinate system to lat/lon
+    if location.coordinate_system == "osgb":
+        longitude, latitude = osgb_to_lon_lat(x=location.x, y=location.y)
+        location = Location(
+            x=longitude,
+            y=latitude,
+            coordinate_system="lon_lat",
+        )
+    elif location.coordinate_system == "geostationary":
+        raise NotImplementedError(
+            "Does not currently support geostationary coordinates when using unstructured grids"
+        )
+
+    # Extract lat, lon, and locidx data
+    lat = xr_data.longitude.values
+    lon = xr_data.latitude.values
+    locidx = xr_data[location_idx_name].values
+
+    # Create a KDTree
+    tree = KDTree(list(zip(lat, lon)))
```

Comment:
what does KDTree do?

Reply:
I'm not sure. This is old code that I just refactored into functions rather than only have it inside a datapipe. I was refactoring the functions I use so I decided just to do the rest